includes/wp-admin/class-admin.php

```diff
@@ -591,20 +596,181 @@ public static function user_bulk_options( $actions ) {
-			'add_activitypub_cap' !== $action
-		) {
-			return $send_back;
+		switch ( $action ) {
+			case 'add_activitypub_cap':
+				foreach ( $users as $user_id ) {
+					$user = new \WP_User( $user_id );
+					$user->add_cap( 'activitypub' );
+				}
+				return $send_back;
+			case 'remove_activitypub_cap':
+				$removed_count = 0;
+
+				// Remove capabilities immediately.
+				foreach ( $users as $user_id ) {
+					$user = new \WP_User( $user_id );
+
+					// Check if user has ActivityPub capability.
+					if ( ! $user->has_cap( 'activitypub' ) ) {
+						continue;
+					}
+
+					// Remove the capability.
+					$user->remove_cap( 'activitypub' );
+
+					// Force cache refresh for user capabilities.
+					\wp_cache_delete( $user_id, 'users' );
+					\wp_cache_delete( $user_id, 'user_meta' );
+
+					++$removed_count;
+				}
+
+				// Build the query args with proper array handling for fediverse deletion confirmation.
+				$query_args = array(
+					'action'    => 'activitypub_confirm_removal',
+					'send_back' => \rawurlencode( $send_back ),
+				);
+
+				// Add user IDs as separate parameters.
+				foreach ( $users as $index => $user_id ) {
+					$query_args[ sprintf( 'users[%d]', $index ) ] = absint( $user_id );
+				}
+
+				$confirmation_url = \add_query_arg( $query_args, \admin_url( 'users.php' ) );
+
+				// Force redirect instead of just returning URL.
+				\wp_safe_redirect( $confirmation_url );
+				exit;
+			case 'delete_actor_confirmed':
+				// Use unified method with no fediverse deletion (keep).
+				return self::process_capability_removal( $users, 'keep', $send_back );
+			default:
+				return $send_back;
 		}
+	}
 
-		foreach ( $users as $user_id ) {
-			$user = new \WP_User( $user_id );
-			if ( 'add_activitypub_cap' === $action ) {
-				$user->add_cap( 'activitypub' );
-			} elseif ( 'remove_activitypub_cap' === $action ) {
-				$user->remove_cap( 'activitypub' );
-			}
+	/**
+	 * Handle the bulk capability removal page request directly.
+	 */
+	public static function handle_bulk_actor_delete_page() {
+		// Check if this is our confirmation page request.
+		// phpcs:ignore WordPress.Security.NonceVerification
+		if ( ! isset( $_GET['action'] ) || 'activitypub_confirm_removal' !== $_GET['action'] ) {
+			return;
+		}
+
+		// Check if we're in admin area.
+		if ( ! \is_admin() ) {
+			return;
+		}
+
+		// Check permissions.
+		if ( ! \current_user_can( 'edit_users' ) ) {
+			\wp_die( \esc_html__( 'You do not have sufficient permissions to access this page.', 'activitypub' ) );
+		}
+
+		// Get parameters.
+		// phpcs:ignore WordPress.Security.NonceVerification, WordPress.Security.ValidatedSanitizedInput
+		$users = \wp_unslash( $_GET['users'] ?? array() );
+		// phpcs:ignore WordPress.Security.NonceVerification
+		$send_back = \urldecode( \sanitize_text_field( \wp_unslash( $_GET['send_back'] ?? '' ) ) );
+
+		// Sanitize user IDs.
+		$users = \array_map( 'absint', (array) $users );
+		$users = \array_filter( $users );
```

Comment:
We should remove users that don't currently have the `activitypub` capability. I was able to get to the confirmation page by selecting any user and running the "Disable for Activitypub" bulk action.